Loadgen: containerize, use terraform to provision and deploy on GKE, CI for image pushing, e2e tests

gzhao408 opened the pull request:
Resolves #404 #427 #428 

-[x] Containerize the load generator
-[x] Provision GKE cluster and deploy load generator using Terraform
-[x] CI workflow: tag and push loadgen docker image on push to master and new release
-[x] Modify/add e2e testing for the load generator
-[x] Remove the old load generator
-[x] Alias kubectl contexts to make it easier fo the user to see what's going on

gzhao408 commented:
[cloud shell link for testing this branch](https://console.cloud.google.com/cloudshell/editor?cloudshell_git_repo=https://github.com/gzhao408/stackdriver-sandbox.git&cloudshell_git_branch=loadgen-gke&cloudshell_working_dir=terraform&shellonly=true&cloudshell_image=gcr.io/stackdriver-sandbox-230822/cloudshell-image:v0.2.5 )

simonz130 commented on kubernetes-config/locust-master-controller.yaml:
```diff
@@ -0,0 +1,52 @@
+    spec:
+      containers:
+        - name: locust-master
+          image: gcr.io/stackdriver-sandbox-230822/sandbox/loadgenerator:wip
```
why this tag?

gzhao408 replied:
just mentioned in the meeting - we'll have it just always use the latest release tag (working with Dan to make this happen)

simonz130 commented on kubernetes-config/locust-master-controller.yaml:
```diff
@@ -0,0 +1,52 @@
+apiVersion: "extensions/v1beta1"
+kind: "Deployment"
+metadata:
+  name: locust-master
```
can we rename locust-master to locust-main or locust-lead or something along these lines? (more inclusive naming)

gzhao408 replied:
sure, will do. potentially more problematic is the Web UI displays "number of slaves" so i'm not super sure how to deal with that

gzhao408 replied:
done

simonz130 commented on terraform/install.sh:
```diff
@@ -198,6 +198,8 @@ applyTerraform() {
   CLUSTER_ZONE=$(gcloud container clusters list --filter="name:cloud-ops-sandbox" --project $project_id --format="value(zone)")
   gcloud container clusters get-credentials cloud-ops-sandbox --zone "$CLUSTER_ZONE"
+  # Make alias for this kubectl context
+  kubectx main=.
```
why do we need this? We don't bundle kubectx into our custom image and I think it's unnecessary dependency

gzhao408 replied:
We need kubectl context to match the cluster (the hipstershop and loadgen are on different clusters). `kubectx` is a pretty easy way to do it (and aliasing it for the user makes a better user exp in my opinion) and @Daniel-Sanche told me the dependency is included in the cloud shell

Daniel-Sanche replied:
Yeah, kubectx is baked into the default Cloud Shell image. I think we should make use of it so we can switch between clusters without needing to find the zone each time

simonz130 replied:
Surprised it's included in base cloud shell image, but ok.

simonz130 commented on terraform/install.sh:
```diff
@@ -256,28 +258,30 @@ getExternalIp() {
-  ../loadgenerator/loadgen autostart $external_ip
+  pushd loadgen/
+  terraform init -lock=false
+  terraform apply --auto-approve -var="project_id=${project_id}" -var="external_ip=${external_ip}"
```
loadgen provisioning should be part of terraform provisioning that includes the rest of the cluster. It should not be invoked separately as this way we'll have a dedicated state (.tfstate) and it will hurt our ability to resume cluster creation in case it didn't work for the first time.

gzhao408 replied:
will work on this - the clusters can be provisioned at the same time, but it'll require a bit of refactoring

simonz130 replied:
Given that your change is already pretty big, I suggest splitting it into a separate CR after this one is merged.

Daniel-Sanche replied:
How much work/time would be involved in making the change?

gzhao408 replied:
i can just put the cluster-provisioning in the same folder as the hipster shop stuff - that is pretty straightforward and achieves our main goal of having those two tasks run concurrently (saving a lot of time). Moving the rest of the load generator stuff over might get a little complicated.

nicklin37 replied:
I'd recommend using a module https://www.terraform.io/docs/configuration/modules.html to be able to keep all the loadgen stuff together and you can pass all the variables you need for the module in. You can use a script to pull the external IP and use the external data Terraform resource to retrieve it for your variables. I have an incomplete version here https://github.com/nicklin37/stackdriver-sandbox/blob/monitoring_integration/terraform/04_monitoring.tf

simonz130 replied:
+1 to Nic's suggestion. We can do it in a new PR though after this one is merged.

simonz130 commented on terraform/install.sh:
```diff
@@ -256,28 +258,30 @@ getExternalIp() {
   # find the IP of the load generator web interface
   TRIES=0
-  while [[ $(curl -sL -w "%{http_code}"  "http://$loadgen_ip:8080" -o /dev/null --max-time 1) -ne 200  && \
+  while [[ $(curl -sL -w "%{http_code}"  "http://$loadgen_ip:8089" -o /dev/null --max-time 1) -ne 200  && \
```
This also can be done via terraform

simonz130 commented on terraform/install.sh:
```diff
@@ -256,28 +258,30 @@ getExternalIp() {
     log "error: load generator unreachable"
   fi
+
+  # Make kubectx alias for this kubectl context
+  kubectx loadgenerator=.
+  # Return kubectl context to the main cluster and show this to the user
+  log $(kubectx main)
```
These should be removed

gzhao408 replied:
i disagree, I think it's best to return the kubectl context to the main one because it's what a user would be more likely to expect. they shouldn't need to really look at what's going on with the load generator.

Daniel-Sanche replied:
Yeah, I also think we should return the user to the main cluster after provisioning the load generator

I'm not so sure about putting it in a log statement though, it would probably be better to do it as two lines to separate the logging from the functionality

simonz130 replied:
Good point Gloria, agree - let's keep it then.

gzhao408 replied:
yeah - i think i'd run `kubectx main` and then log `kubectx` to print the current context?

simonz130 commented on terraform/loadgen/00_loadgen.tf:
```diff
@@ -0,0 +1,150 @@
+# zone in us-west1 at random - a different region from where we provisioned
+# the Hipster Shop.
+resource "random_shuffle" "zone" {
+  input = ["us-west1-a", "us-west1-b", "us-west1-c"]
```
add other locations that are not us-west and are not locations for creation of the demo app cluster

gzhao408 replied:
afaik the demo app cluster only uses `us-central1`, and loadgen cluster only uses `us-west1`... is it necessary to have a lot of possible regions? as long as they're in different geographic areas?

simonz130 replied:
yeah, different geo-locations is important. Having it be based not only in US is a secondary but might benefit folks too

gzhao408 replied:
yeah I could do that! wasn't sure if it'd be restricted or something, but i can add a few other regions

gzhao408 replied:
just added some, our loadgenerator is Mr. Worldwide

simonz130 commented on terraform/loadgen/00_loadgen.tf:
```diff
@@ -0,0 +1,150 @@
+
+  # Stores the zone of created loadgenerator gke cluster
+  provisioner "local-exec" {
+    command = "gcloud config set compute/zone ${element(random_shuffle.zone.result, 0)}"
```
once you integrate this with main terraform script, you might need to change this so the zone is appended for each gcloud command that needs it as opposed to setting it up globally like this

simonz130 commented on terraform/loadgen/00_loadgen.tf:
```diff
@@ -0,0 +1,150 @@
+# Setting kubectl context to currently deployed loadgenerator GKE cluster
+resource "null_resource" "set_gke_context" {
+  provisioner "local-exec" {
+    command = "gcloud container clusters get-credentials cloud-ops-sandbox-loadgenerator --zone ${element(random_shuffle.zone.result, 0)} --project ${var.project_id}"
```
use variable substitution that evaluates to cloud-ops-sandbox-loadgenerator (instead of using a literal constant)

gzhao408 replied:
it's not variable though, we name it in the terraform script? could put it in a constant i guess, but we do the exact same for cloud-ops-sandbox above

simonz130 commented on tests/provisioning/loadgen_test.py:
```diff
@@ -0,0 +1,69 @@
+        command = ("kubectl get service locust-master -o jsonpath='{.status.loadBalancer.ingress[0].ip}'")
+        result = subprocess.run(split(command), encoding='utf-8', capture_output=True)
+        loadgen_ip = result.stdout.replace('\n', '')
+        url = 'http://{0}:8089'.format(loadgen_ip)
```
extract port into const

gzhao408 replied:
done

simonz130 commented on tests/provisioning/loadgen_test.py:
```diff
@@ -0,0 +1,69 @@
+        self.assertTrue(urllib.request.urlopen(url).getcode() == 200)
+
+    def testDifferentZone(self):
+        """ Test if load generator is in a different zone from the GKE cluster """
```
...from the demo app cluster

simonz130 commented on .github/workflows/push-master.yml:
```diff
@@ -36,3 +36,18 @@ jobs:
+        set -x
+        gcloud auth configure-docker --quiet
+        # build and tag with git hash
+        IMAGE_SHA=gcr.io/$PROJECT_ID/sandbox/loadgenerator:$GITHUB_SHA
+        docker build -t $IMAGE_SHA ./src/loadgenerator
```
Should we have a different path here to reflect new loadgen changes compared to previous image?

gzhao408 replied:
oop! let me fix this up, slipped my mind

Daniel-Sanche commented on terraform/install.sh:
```diff
@@ -256,28 +258,30 @@ getExternalIp() {
     TRIES=$((TRIES + 1))
   done
-  if [[ $(curl -sL -w "%{http_code}"  "http://$loadgen_ip:8080" -o /dev/null  --max-time 1) -ne 200 ]]; then
+  if [[ $(curl -sL -w "%{http_code}"  "http://$loadgen_ip:$LOCUST_PORT" -o /dev/null  --max-time 1) -ne 200 ]]; then
```
nit: I don't like seeing the query duplicated here. It would be great if you could find a way to use the result from above. But I know your time is limited, so don't worry too much about this

Daniel-Sanche commented on terraform/install.sh:
```diff
@@ -256,28 +258,30 @@ getExternalIp() {
-                                               --format="value(networkInterfaces[0].accessConfigs.natIP)")
+    sleep 10
+    loadgen_ip=$(kubectl get service locust-main -o jsonpath='{.status.loadBalancer.ingress[0].ip}');
+     [ -z "$loadgen_ip" ] && sleep 10;
```
is sleeping both before and after the `kubectl` line intentional?

nicklin37 replied:
+1 I think if you're waiting for the service already in the Terraform it might be a bit extra to wait twice here.

gzhao408 replied:
this was causing the bug haha, it's part of the while loop.
this is waiting for the frontend to become available, not the service to be running

Daniel-Sanche replied:
That's a bit concerning. Instead of generic sleeps, maybe you could try a `kubectl wait` for the frontend to be ready? (But this is low priority at this point)

Daniel-Sanche commented on terraform/loadgen/00_loadgen.tf:
```diff
@@ -0,0 +1,150 @@
+  # interesting things.
+  node_pool {
+    node_config {
+      machine_type = "n1-standard-2"
```
I'm a bit unsure about needing 3 n1-standard-2 nodes. That could get expensive if customers accidently leave them live, and we really shouldn't need too many resources for a simple load generator

gzhao408 replied:
lowered to 1

simonz130 commented on tests/provisioning/loadgen_test.py:
```diff
@@ -0,0 +1,69 @@
+        command=('gcloud config set project {0}'.format(getProjectId()))
+        subprocess.run(split(command))
+        # set kubectl context to cloud-ops-sandbox-loadgenerator
+        command=('gcloud container clusters get-credentials cloud-ops-sandbox-loadgenerator --zone {0}'.format(getClusterZone()))
```
This might be problem when your run e2e tests of various components involving kubernetes on the same CI agent.
For example, say you run loadgen e2e test wired against gke1 and hipstershop e2e test wired to gke2.
If you switch the context, it's switches it for all the sessions for the same user on the same VM/container, so by switching to gke1, you might affect tests that run in parallel and expect to run against gke2.
To prevent that, in your tests, you should specify _--context_ with every kubectl command.

simonz130 commented on tests/provisioning/provision_test.py:
```diff
@@ -30,6 +30,12 @@ class TestGKECluster(unittest.TestCase):
+        command=('gcloud config set project {0}'.format(getProjectId()))
+        subprocess.run(split(command))
+        # set kubectl context
+        command=('gcloud container clusters get-credentials cloud-ops-sandbox --zone {0}'.format(getClusterZone()))
```
see a similar comment in loadgen tests

gzhao408 commented on .github/workflows/push-master.yml:
```diff
@@ -36,3 +36,18 @@ jobs:
+        docker build -t $IMAGE_SHA ./src/loadgenerator
+        docker push $IMAGE_SHA
+        # build and tag with latest
+        IMAGE_LATEST=gcr.io/$PROJECT_ID/sandbox/loadgenerator:latest
```
just pushed a change, renamed to [gcr.io/stackdriver-sandbox-230822/sandbox/loadgenerator/basic:latest](https://pantheon.corp.google.com/gcr/images/stackdriver-sandbox-230822/GLOBAL/sandbox/loadgenerator/basic?gcrImageListsize=30)

gzhao408 commented on kubernetes-config/locust-worker-controller.yaml:
```diff
@@ -0,0 +1,43 @@
+  labels:
+    name: locust-worker
+spec:
+  replicas: 5
```
oops 😅  mostly just followed by example. what would you recommend for number of workers, type of machine, etc? I have no concept of what we'd need tbh

Daniel-Sanche replied:
I usually do some trial and error. I can't imagine we need more than 1 worker given our cap is 150QPS. I wonder if it's possible/feasible to use a single container instead of a master/workers (I thought that's what we did on upstream hipstershop)

For machine types, it might be good to see if a single n1-standard-2 works. if not, maybe 2 n1-standard-1's? Or try out a single higher-CPU node. We want to use the minimum resources so we're not wasting customer's money, but you don't need to spend too much time experimenting

gzhao408 replied:
Ok I basically just reduced everything to 1 and things still work. From my humble noobernetes + locust knowledge, apparently the worker is there for scalability (we can increase the number of replicas / locust workers easily?)

Daniel-Sanche replied:
Ok sounds good! I don't think we need to worry too much about load generator scalability when the service itself only handles 150QPS, but yeah it's good to have it set up for easy growth later

simonz130 replied:
The service itself handles 150qps - this needs to be fixed (not in this PR). The scalability of the loadgen will help us to ramp-up the number and variety of load patterns we can send into the cluster.

gzhao408 commented:
just pushed a bunch of small fixes but I tested it and I think I accidentally messed up the install script. a few moments please...

gzhao408 commented on kubernetes-config/locust-main-service.yaml:
```diff
@@ -0,0 +1,38 @@
+    app: locust-main
+spec:
+  ports:
+    - port: 8089
```
I think it's locust's default port for whatever reason 🤷  I can switch it out in a followup

Daniel-Sanche replied:
Using the service, you can map any external port number to any internal one. So locust can keep listening at 8089, but you can expose it to users at the default :80 if you want

But this is pretty low priority, just a suggestion

gzhao408 commented on terraform/install.sh:
```diff
@@ -14,7 +14,7 @@
 
 # This script provisions Hipster Shop Cluster for Cloud Operations Sandbox using Terraform
-
+LOCUST_PORT="8089"
```
moved it back down, yeah the global up top doesn't make much sense

gzhao408 commented:
Ok everything works now. Last changes:
- The image used in config files and auto-pushed in CI is loadgenerator/basic:latest. This doesn't interfere with any current loadgen stuff.
- Fewer resources, 1 worker replica, 1 n1-standard-2.
- Load Generator Worldwide™
- Cleaned up the install script - use non-global constants, don't log and execute at the same time, check variable assignment instead of repeating curl command
- Rename locust-master to locust-main
- Track the provisioning test_runner in git (oops)
- docs/style/nits

@Daniel-Sanche Ready 4 Merj? 😄

Daniel-Sanche reviewed (approved):
LGTM!

Daniel-Sanche commented on make-release.sh:
```diff
@@ -38,6 +38,9 @@ sed -i -e "s/cloudshell-image:v\([0-9\.]\+\)/cloudshell-image:${NEW_VERSION}/g"
 sed -i -e "s/cloudshell-image:v\([0-9\.]\+\)/cloudshell-image:${NEW_VERSION}/g" ${REPO_ROOT}/docs/index.html;
 
+# update tags for load generator image
+find "${REPO_ROOT}/kubernetes-config" -name '*.yaml' -exec sed -i -e "s/basic:[0-9]+\.[0-9]+\.[0-9]+/basic:${NEW_VERSION}/g" {};
```
This seems to be doing a find and repace on a version code, when it should be replacing latest
You'll also need to add a line to revert back to latest later, and to run git add on the file

You can run `./make-release --dry-run` and  `./make-release --no-push` to make sure everything works as expected

Daniel-Sanche commented on make-release.sh:
```diff
@@ -38,6 +38,9 @@ sed -i -e "s/cloudshell-image:v\([0-9\.]\+\)/cloudshell-image:${NEW_VERSION}/g"
 sed -i -e "s/cloudshell-image:v\([0-9\.]\+\)/cloudshell-image:${NEW_VERSION}/g" ${REPO_ROOT}/docs/index.html;
 
+# update tags for load generator image
+find "${REPO_ROOT}/kubernetes-config" -name '*.yaml' -exec sed -i -e "s/:latest/:${NEW_VERSION}/g" {};
```
This looks like it's missing the \ at the end. Otherwise it gives an error

Also, it would probably fit better with the other line under `# temporarily pin manifests to :$NEW_VERSION`

Daniel-Sanche commented on make-release.sh:
```diff
@@ -56,7 +59,10 @@ else
+    git commit -m "revert demo app images to latest"
+    find "${REPO_ROOT}/kubernetes-config" -name '*.yaml' -exec sed -i -e "s/:${NEW_VERSION}/:latest/g" {} \;
+    git add "${REPO_ROOT}/kubernetes-config/*.yaml"
+    git commit -m "revert loadgeenerator images to latest"
```
This shouldn't have to be a new commit. You should be able to make the changes as part of the other one

Also, you need to make the `git add "${REPO_ROOT}/kubernetes-config/*.yaml"` part of the other commit too, to make sure it's in the release commit (put it under `# create release commit`)

Daniel-Sanche replied:
Also, this is a larger problem I'm just noticing now: Why is the load generator in a new `/kubernetes-config` file? It's confusing to have both that and `/kubernetes-manifests` in the repo root. i'd suggest putting your manifests in `/kubernetes-manifests/load-generator`

gzhao408 commented:
refactored the manifests, fixed up some stuff in make-release script

i added this branch to the e2e config so it'll run here, also testing using the cloud shell link.
once it passes i'll revert the top commit

gzhao408 commented:
Reverted https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/pull/429/commits/0a27a520ed162af085491e030def81fc0b19ec1c, it passed.

Daniel-Sanche reviewed (approved):
LGTM

gzhao408 commented:
OK IT WORKS
